[Static graph] Changes to enable building roslyn

cdmihai opened the pull request:
Things needed to build `roslyn/Compilers.sln` with no changes to the roslyn repo.

cdmihai commented:
cc @AndyGerlicher @smera

rainersigwald commented on src/Build/Graph/ProjectGraphNode.cs:
```diff
@@ -35,12 +39,34 @@ internal ProjectGraphNode(ProjectInstance projectInstance)
+        {
+            foreach (var reference in _projectReferences)
+            {
+                ErrorUtilities.VerifyThrow(reference._referencingProjects.Contains(this), "references should point the nodes referencing them");
```
```suggestion
                ErrorUtilities.VerifyThrow(reference._referencingProjects.Contains(this), "references should point to the nodes referencing them");
```

rainersigwald commented on src/Build/Graph/GraphBuilder.cs:
```diff
@@ -0,0 +1,397 @@
+        internal class GraphEdges
+        {
+            private ConcurrentDictionary<(ProjectGraphNode, ProjectGraphNode), ProjectItemInstance> ReferenceItems =
+                new ConcurrentDictionary<(ProjectGraphNode, ProjectGraphNode), ProjectItemInstance>();
```
This data type raises some concerns for me. Can you expand on why it's needed?

cdmihai replied:
The change requires keeping extra state for each edge (the project reference metadata for each reference). There's two ways to keep it:
1. Enhance the OM by adding a GraphEdge class, and then have ProjectGraphNode reference lists of GraphEdge objects, rather than directly referencing parent/child ProjectGraphNodes
2. Hack it out by introducing a dictionary.

1. implies either changing the public API to add GraphEdge, or making it private by adding a dictionary on each node, which is memory wasteful. Since it's not clear whether anyone would find the extra public API useful, I decided to just go with 2. and keep everything private. If, in the future, somebody needs to access this data, then we can spend more energy in designing some cute looking API.

livarcocc commented on src/Build/Graph/GraphBuilder.cs:
```diff
@@ -0,0 +1,397 @@
+
+            var allParsedProjects = new ConcurrentDictionary<ConfigurationMetadata, ProjectGraphNode>();
+            var tasksInProgress = new ConcurrentDictionary<ConfigurationMetadata, object>();
+            var projectsToEvaluate = new ConcurrentQueue<ConfigurationMetadata>();
```
Why not just pass the entryPointConfigurationMetadata to the ConcurrentQueue constructor? It is a List and therefore an IEnumerable. If you do that you won't need the foreach below.

rainersigwald commented on src/Build/Graph/ProjectGraph.cs:
```diff
@@ -573,6 +530,20 @@ private static ImmutableList<string> ExpandDefaultTargets(ImmutableList<string>
+                        .RemoveAt(i)
+                        .InsertRange(i, expandedTargets);
+
+                    i += defaultTargets.Count;
```
`defaultTargets` or `expandedTargets`?

cdmihai replied:
Good catch! Fortunately the defect does not affect correctness.

rainersigwald commented on src/Build/Graph/ProjectInterpretation.cs:
```diff
@@ -115,7 +115,12 @@ internal static ProjectType GetProjectType(ProjectInstance project)
+        /// <summary>
+        /// To avoid calling nuget at graph construction time, the graph is initially constructed with outer build nodes referencing inner build nodes.
+        /// However, at build time, for non root outer builds, the inner builds are NOT referenced by the outer build, but by the nodes referencing the
+        /// outer build. Change the graph to mimic this behaviour.
```
I don't understand. It sounds like "it starts outer->inner, but at build time it's actually outer->inner", but that doesn't make sense. Can you rephrase?

cdmihai replied:
Updated comment with example.

livarcocc commented on src/Build/Graph/GraphBuilder.cs:
```diff
@@ -0,0 +1,397 @@
+
+            IReadOnlyCollection<ProjectGraphNode> GetGraphRoots(IReadOnlyCollection<ProjectGraphNode> entryPointNodes)
+            {
+                var graphRoots = new List<ProjectGraphNode>(entryPointNodes.Count);
```
Ok to waste memory here if not all entryPoints end up being root?

cdmihai replied:
Trimmed the excess memory.

livarcocc commented on src/Build/Graph/GraphBuilder.cs:
```diff
@@ -0,0 +1,397 @@
+                    var configurationMetadata = new ConfigurationMetadata(FileUtilities.NormalizePath(entryPoint.ProjectFile), globalPropertyDictionary);
+                    entryPointConfigurationMetadata.Add(configurationMetadata);
+                }
+                return entryPointConfigurationMetadata;
```
NIT: empty line after closing bracket.

rainersigwald commented on src/Build.UnitTests/Graph/ProjectGraph_Tests.cs:
```diff
@@ -769,6 +851,78 @@ public void GetTargetListsFiltersTargetsForOuterAndInnerBuilds()
         }
 
+        [Fact]
+        public void GetTargetListsDoesNotUseTargetsMetadataOnInnerBuildsFromRootOuterBuilds()
```
May I beg you for a diagram of this?

cdmihai replied:
![image](https://user-images.githubusercontent.com/2255729/58132013-10b41a80-7bd5-11e9-9c92-0b192b112416.png)

livarcocc commented on src/Build/Graph/GraphBuilder.cs:
```diff
@@ -0,0 +1,397 @@
+        ///     Maintain a queue of projects to be processed and evaluate projects in parallel
+        ///     Returns false if loading the graph is not successful
+        /// </summary>
+        private bool FindGraphNodes(
```
Should this be LoadGraphNodes instead? Find seems to imply a search and I would not expect modifications.

cdmihai replied:
It finds and loads them, but does not connect them. What do you think about `FindAndLoadGraphNodes`? Seems a bit superfluous.

rainersigwald commented on src/Build.UnitTests/Graph/ProjectGraph_Tests.cs:
```diff
@@ -1239,6 +1419,13 @@ private static void AssertOuterBuildAsNonRoot(ProjectGraphNode outerBuild, Dicti
                     innerBuild.ReferencingProjects.ShouldContain(outerBuildReferencer);
                     innerBuild.ReferencingProjects.ShouldNotContain(outerBuild);
+
+                    graph.TestOnly_Edges.TestOnly_HasEdge((outerBuild, innerBuild)).ShouldBeFalse();
```
Wait, why no edge for outer->inner of the same project?

cdmihai replied:
These are outer builds that are not roots. For non-root outerbuilds, the project that references the outer build also references the inner builds.

rainersigwald commented on src/Build/Graph/ProjectGraphNode.cs:
```diff
@@ -10,6 +11,7 @@ namespace Microsoft.Build.Experimental.Graph
     /// <summary>
     /// Represents the node for a particular project in a project graph.
     /// </summary>
+    [DebuggerDisplay(@"{ToString()}")]
```
You could do this without exposing `ToString()` as public API by calling the method something else; might be worth doing.

cdmihai commented:
> Can you lay out what the problems were when building Roslyn that you've resolved here? That would help place the review.
> 
> The commit `Read ProjectReferenceItem.Targets` is not landing for me clearly. I think I understand the goal but not the implementation.

Some Roslyn projects set the `Targets` metadata on certain `ProjectReference` items. Static graph did not initially take this metadata into account, and therefore it inferred the wrong entry targets for the references which had that metadata set. So then, in addition to the `.default` symbol I also added `.projectReferenceTargetsOrDefault` and updated the target protocol to use it instead of `.default`.

The actual change implies remembering the project reference metadata from graph construction, so it can be read by `ProjectGraph.GetTargetLists` when it infers entry targets.

cdmihai commented on src/Build/Graph/GraphBuilder.cs:
```diff
@@ -0,0 +1,397 @@
+        ///     Maintain a queue of projects to be processed and evaluate projects in parallel
+        ///     Returns false if loading the graph is not successful
+        /// </summary>
+        private bool FindGraphNodes(
```
Good question. Moved it inside the method.

cdmihai commented:
@livarcocc @rainersigwald responded to all feedback
